source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+    case SpvOp::SpvOpINotEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() != operand_words.back();
+    case SpvOp::SpvOpULessThan:
+      assert(operand_words.size() == 2);
+      return operand_words.front() < operand_words.back();
+    case SpvOp::SpvOpSLessThan:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) < ((int32_t)operand_words.back());
+    case SpvOp::SpvOpUGreaterThan:
+      assert(operand_words.size() == 2);
+      return operand_words.front() > operand_words.back();
+    case SpvOp::SpvOpSGreaterThan:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) > ((int32_t)operand_words.back());
+    case SpvOp::SpvOpULessThanEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() <= operand_words.back();
+    case SpvOp::SpvOpSLessThanEqual:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) <=
+             ((int32_t)operand_words.back());
+    case SpvOp::SpvOpUGreaterThanEqual:
+      assert(operand_words.size() == 2);
+      return operand_words.front() >= operand_words.back();
+    case SpvOp::SpvOpSGreaterThanEqual:
+      assert(operand_words.size() == 2);
+      return ((int32_t)operand_words.front()) >=
+             ((int32_t)operand_words.back());
+    default:
+      assert(false);
+      return 0;
+  }
+}
+
+// A helper function to check the validity for broadcast operations. Returns
+// true if all of the given operands are:
+//  1) vector type front-end constants;
+//  2) same in the number of components.
+// This method is only used in assertions for now.
+bool AllInOperandValuesAreConstVectorsInSameLength(
+    ir::Instruction* inst,
+    const std::unordered_map<uint32_t, std::unique_ptr<Value>>&
+        id_to_const_val) {
+  size_t len = 0;
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    if (inst->GetInOperand(i).type == SPV_OPERAND_TYPE_ID) {
+      uint32_t id = inst->GetSingleWordInOperand(i);
+      // Should have a value recorded for each operand.
+      auto iter = id_to_const_val.find(id);
+      if (iter == id_to_const_val.end()) return false;
+      // The recorded values for all operands should be in composite type.
+      if (CompositeValue* val = iter->second->AsComposite()) {
+        // The recorded values for all operands should have same number of
+        // components.
+        len = len == 0 ? val->nested.size() : len;
+        if (len != val->nested.size()) return false;
+        // Each component of an operand should be a scalar, so the record value
+        // should also be in scalar type.
+        for (uint32_t nested_id : val->nested) {
+          auto nested_iter = id_to_const_val.find(nested_id);
+          if (nested_iter == id_to_const_val.end()) return false;
+          if (!nested_iter->second->AsScalar()) return false;
+        }
+      } else {
```

Comment:
This comment is just my personal habit, so your choice.

For this kind of one-liner `return` code block, I typically have it dealt at the very beginning of the enclosing block, i.e., `if (inst->GetInOperand(i).type == SPV_OPERAND_TYPE_ID) return false;` I find this makes the code easier to read since I don't need to match the indent/brackets to find which `if` this block is associated. And the bigger code block (the current `if` body) is not indented yet another level.

But, pure personal attitude. Your choice.